### PR TITLE
Bug 1764313: 4.3 console_url is no longer reported on metrics endpoint (fix, with e2e test)

### PR DESCRIPTION
**Pull request by benjaminapetersen**
An e2e test to validate that the `/metrics` endpoint on the `operator` responds with an expected output.

### CODE REVIEW

**Comment by jhadvig** [comment]
> Currently will not run due to a 403 error on the endpoint. The test suite is not allowed to hit the endpoint. Needs to decide the best solution for this.

Is this due to the missing operator/operand implementation ?

**Comment by benjaminapetersen** [comment]
I'm not really sure. I'm assuming we are running the tests as `kube:admin` since we use our `kubeconfig`.  So I would expect we should be allowed to make a request against the `/metrics` url, but I get `403 Forbidden`.

File: test/e2e/metrics_test.go [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+}
+
+// gets us a token from the kubeconfig file
+func getBearerToken(t *testing.T) string {
+	tokenProvider, err := framework.GetConfig()
+	if err != nil {
+		t.Fatalf("error, can't get config: %s", err)
+	}
+	// build the request header with a token from the kubeconfig
+	return fmt.Sprintf("Bearer %s", tokenProvider.BearerToken)
```
Comment (benjaminapetersen) [inline test/e2e/metrics_test.go]: Getting the `token` from the `kubeconfig` made the test possible.

File: test/e2e/metrics_test.go [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+	if err != nil {
+		t.Fatalf("error: %s", err)
+	}
+	t.Logf("route to /metrics: (%v) \n", routeForMetrics)
+	return routeForMetrics
+}
+
+// In production our metrics endpoint should not have a route, but this makes it
+// easier to access the pod http://localhost:8443/metrics endpoint to verify its output
+func tempRouteForTesting() *routev1.Route {
```
Comment (benjaminapetersen) [inline test/e2e/metrics_test.go]: Short term route, just to make the request.

**Comment by benjaminapetersen** [comment]
Ah, that is unfortunate as I know the `release-4.2` will work.  At present it appears our `console_url` metric died in `4.3`. Exploring.

File: test/e2e/metrics_test.go [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+func getBearerToken(t *testing.T) string {
+	tokenProvider, err := framework.GetConfig()
+	if err != nil {
+		t.Fatalf("error, can't get config: %s", err)
+	}
+	// build the request header with a token from the kubeconfig
+	return fmt.Sprintf("Bearer %s", tokenProvider.BearerToken)
+}
+
+// creates a temp route, polls for the route, and returns just a formatted url for https://<host>/metrics
```
Comment (jhadvig) [inline test/e2e/metrics_test.go]: route is created in the setup func

File: test/e2e/metrics_test.go [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+			return false, err
+		}
+		routeForMetrics = "https://" + tempRoute.Spec.Host + "/metrics"
+		t.Logf("route for metrics: %v", routeForMetrics)
+		return true, nil
+	})
+	if err != nil {
+		t.Fatalf("error: %s", err)
+	}
+	t.Logf("route to /metrics: (%v) \n", routeForMetrics)
```
Comment (jhadvig) [inline test/e2e/metrics_test.go]:
```suggestion
	t.Logf("route to /metrics: (%s) \n", routeForMetrics)
```

File: test/e2e/metrics_test.go [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+// 	curl --insecure -H "Authorization: Bearer $(oc whoami --show-token)" $(oc get route metrics --template 'https://{{.spec.host}}/metrics') | grep console_url
+// alternatively:
+// 	oc exec -it console-operator-<pod-id> /bin/bash
+// 	curl --insecure -H "Authorization: Bearer <token>  https://localhost:8443/metrics | grep console_url
+func TestMetricsEndpoint(t *testing.T) {
+	client, _ := setupMetricsEndpointTestCase(t)
+	defer cleanUpMetricsEndpointTestCase(t, client)
+
+	metricsURL := getMetricsURL(t, client)
+	fmt.Printf("fetching metrics url.... %v\n", metricsURL)
```
Comment (jhadvig) [inline test/e2e/metrics_test.go]: nit: should be before the `getMetricsURL()`

File: test/e2e/metrics_test.go [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+
+	fmt.Printf("making request...\n")
+	respString := metricsRequest(t, metricsURL)
+
+	toFind := "console_url"
+	fmt.Printf("searching for %v\n", toFind)
```
Comment (jhadvig) [inline test/e2e/metrics_test.go]:
```suggestion
	fmt.Printf("searching for %s\n", toFind)
```

File: test/e2e/metrics_test.go [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+	found := findLineInResponse(respString, toFind)
+	if !found {
+		t.Fatalf("did not find %v", toFind)
```
Comment (jhadvig) [inline test/e2e/metrics_test.go]:
```suggestion
		t.Fatalf("did not find %s", toFind)
```

File: test/e2e/metrics_test.go [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+	}
+}
+
+func findLineInResponse(haystack, needle string) (found bool) {
+	scanner := bufio.NewScanner(strings.NewReader(haystack))
+	for scanner.Scan() {
+		found := strings.Contains(scanner.Text(), needle)
+		if found {
+			fmt.Printf("found %v\n", scanner.Text())
```
Comment (jhadvig) [inline test/e2e/metrics_test.go]:
```suggestion
			fmt.Printf("found %s\n", scanner.Text())
```

File: test/e2e/metrics_test.go [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+	// ignore self signed certs for testing purposes
+	insecureTransport := &http.Transport{
+		TLSClientConfig: &tls.Config{
+			InsecureSkipVerify: true,
+		},
+	}
+	insecureClient := &http.Client{
+		Transport: insecureTransport,
+	}
+	return insecureClient
```
Comment (jhadvig) [inline test/e2e/metrics_test.go]: lets make it a single struct
```suggestion
	return &http.Client{
		Transport: http.Transport{
			TLSClientConfig: tls.Config{
				InsecureSkipVerify: true,
			},
		},
	}
```

File: test/e2e/metrics_test.go [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+	err := wait.Poll(1*time.Second, 30*time.Second, func() (stop bool, err error) {
+		tempRoute, err := client.Routes.Routes(api.OpenShiftConsoleOperatorNamespace).Get(tempRoute.Name, metav1.GetOptions{})
+		if err != nil {
+			t.Fatalf("error: %s", err)
+		}
+		if len(tempRoute.Spec.Host) == 0 {
+			return false, err
+		}
+		routeForMetrics = "https://" + tempRoute.Spec.Host + "/metrics"
+		t.Logf("route for metrics: %v", routeForMetrics)
```
Comment (jhadvig) [inline test/e2e/metrics_test.go]:
```suggestion
		t.Logf("route for metrics: %s", routeForMetrics)
```

File: test/e2e/metrics_test.go [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+	client, _ := setupMetricsEndpointTestCase(t)
+	defer cleanUpMetricsEndpointTestCase(t, client)
+
+	metricsURL := getMetricsURL(t, client)
+	fmt.Printf("fetching metrics url.... %v\n", metricsURL)
+
+	fmt.Printf("making request...\n")
+	respString := metricsRequest(t, metricsURL)
+
+	toFind := "console_url"
```
Comment (jhadvig) [inline test/e2e/metrics_test.go]: Not sure if this shouldn't be a const after all the imports.

Also would make it a bit more clear that the string is just a metadata field, maybe!?:
```suggestion
const (
	metadataToFind = "console_url"
)
```

File: test/e2e/metrics_test.go (hunk 5 above)
Comment (benjaminapetersen) [reply test/e2e/metrics_test.go]: I do want to print the `%s\n", metricsURL`, however.

File: test/e2e/metrics_test.go (hunk 11 above)
Comment (benjaminapetersen) [reply test/e2e/metrics_test.go]: Sounds good!

File: test/e2e/metrics_test.go (hunk 7 above)
Comment (benjaminapetersen) [reply test/e2e/metrics_test.go]: will do.

File: test/e2e/metrics_test.go (hunk 9 above)
Comment (benjaminapetersen) [reply test/e2e/metrics_test.go]: and `return` immediately, done.

File: test/e2e/metrics_test.go (hunk 3 above)
Comment (benjaminapetersen) [reply test/e2e/metrics_test.go]: will update.

File: test/e2e/metrics_test.go (hunk 5 above)
Comment (benjaminapetersen) [inline test/e2e/metrics_test.go]: `t.Logf` for consistency.

**Comment by benjaminapetersen** [comment]
/hold

Now just needs PR to fix the missing `console_url` (separate, as this test is going to be cherry-picked to `4.2` & `4.1`.

**Comment by benjaminapetersen** [comment]
Swapping to `k8s.io` `legacyregistry` for metrics gets our metrics back (not sure why the `[ALPHA]`):

```
# HELP authentication_attempts [ALPHA] Counter of authenticated attempts.
# TYPE authentication_attempts counter
authentication_attempts{result="success"} 9
# HELP console_url [ALPHA] URL of the console exposed on the cluster
# TYPE console_url gauge
console_url{url="https://console-openshift-console.apps.bpetersen.devcluster.openshift.com"} 1
# HELP go_gc_duration_seconds A summary of the GC invocation durations.
```
nice assist by @jmrodri finding `k8s.io/component-base/metrics/legacyregistry`.

**Comment by benjaminapetersen** [comment]
Leaving this as separate commits so we can `cherry-pick` the e2e test & the delayed registry if we like.  The `legacyregistry` change is new as of a may 2019 commit `move global registry code into subdirectory 'legacyregistry'`.

**Comment by benjaminapetersen** [comment]
/retest 

`
error: unable to extract layer sha256:6bacebbd64e0dd9af09340d762fddc67562c410dac8219efec5d04c388c1242f from registry.svc.ci.openshift.org/ci-op-hv4l9dnw/stable-initial@sha256:98cabd736c67dafd9e14c876c037ff35f01b5b91feadf8f7d1b5095fcbac05b7: stream error: stream ID 53; INTERNAL_ERROR
`

**Comment by benjaminapetersen** [comment]
/retest

? flake? works locally. debugging.

**Comment by benjaminapetersen** [comment]
/retest

`could not wait for build: the build src failed after 35s with reason DockerBuildFailed: Docker build strategy has failed.`

**Comment by benjaminapetersen** [comment]
/retest

`console` tests fail.

File: test/e2e/metrics_test.go [hunk 12]
```diff
@@ -118,16 +118,69 @@ func httpOK(resp *http.Response) bool {
...
+	}
+
+	if err != nil {
+		// not sure if we should panic/die here
+		fmt.Printf("x509 certs err: %v", err)
+	}
+
+	// need to add the CAData, the kubeconfig has router certs
+	// that are missing from the system trust roots
+	rootCAs.AppendCertsFromPEM(config.CAData)
```
Comment (benjaminapetersen) [inline test/e2e/metrics_test.go]: @stlaz thanks for your help here!

**Comment by benjaminapetersen** [comment]
+1 sounds good!

**Comment by benjaminapetersen** [comment]
@jhadvig squashed into 3 commits:

- Delay registry of console_url metric until starter.go
  - this is more of a best practices deal and may be part of what we backport.
- Use k8s.io legacyregistry for metrics
  - this is the fix for this missing metrics in 4.3 and cannot be part of the backport.
- Add e2e test for /metrics endpoint 
  - this is the e2e test and can be part of the backport.

Backport will simply require a cherry pick of 3 commits:
- initial addition of `/metrics` from #270, commit `346399b678`
- delay registry (probably optional)  from this PR #299 commit `d5fe6fedc`
- e2e test from this PR #299 commit `656a8b0913`

I will create a branch for the backport to 4.1 once this merges and we verify our 4.3 metrics are again happy (backport will then require the backport for telemeter as well)

**Comment by benjaminapetersen** [comment]
/retest

almost

**Comment by benjaminapetersen** [comment]
@jhadvig looks like we are good to merge!

**Comment by benjaminapetersen** [comment]
@spadgett care to tag?

File: test/e2e/metrics_test.go [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+	}
+	framework.StandardCleanup(t, client)
+}
+
+// This test essentially does the following to verify the `console_url` is in the metrics endpoint:
+// 	curl --insecure -H "Authorization: Bearer $(oc whoami --show-token)" $(oc get route metrics --template 'https://{{.spec.host}}/metrics') | grep console_url
+// alternatively:
+// 	oc exec -it console-operator-<pod-id> /bin/bash
+// 	curl --insecure -H "Authorization: Bearer <token>  https://localhost:8443/metrics | grep console_url
+func TestMetricsEndpoint(t *testing.T) {
```
Comment (spadgett) [inline test/e2e/metrics_test.go]: 👍

**Comment by benjaminapetersen** [comment]
lol forgot there was a hold on this 😄

**Pull request merged**